drivers: i2c: support for Nuvoton numaker I2C

cyliangtw opened the pull request:
This PR is to add Nuvoton numaker series I2C controller, to support dual role master & slave.
It's verified on board: NuMaker-PFM-M467.

fabiobaltieri commented on drivers/i2c/i2c_numaker.c:
```diff
@@ -0,0 +1,808 @@
+ * 2. Do data transfer via above callback (ISR)
+ */
+
+static void m_numaker_i2c_master_xfer_msg_read_next_byte(const struct device *dev);
+static void m_numaker_i2c_master_xfer_msg_read_last_byte(const struct device *dev);
+static void m_numaker_i2c_master_xfer_msg_end(const struct device *dev);
+static void m_numaker_i2c_master_xfer_end(const struct device *dev, bool do_stop);
```
could you move those functions up and avoid having to forward declare them? It'd also be nice to have `DT_INST_FOREACH_STATUS_OKAY` last

cyliangtw replied:
of course, I can follow it.

fabiobaltieri commented on drivers/i2c/i2c_numaker.c:
```diff
@@ -0,0 +1,808 @@
+	}
+
+	/* Configure pinmux (NuMaker's SYS MFP) */
+#ifdef CONFIG_PINCTRL
+	err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (err != 0) {
+		goto cleanup;
+	}
```
I think you should just `select PINCTRL` and then take it for granted that it's there, that's how nrfx does and I'd say how other platforms should do it as well, no sense failing the build just so that the user have to enable the option manually when the system can select it for them

cyliangtw replied:
Sorry, this is the old-tyle compatibility code for the proprietary pin-ctl. This line was omitted during the upgrade.
Now, fixed it.

fabiobaltieri commented on drivers/i2c/i2c_numaker.c:
```diff
@@ -0,0 +1,808 @@
+		goto cleanup;
+	}
+	/* Equivalent to CLK_SetModuleClock() */
+	err = clock_control_configure(config->clkctrl_dev,
+		(clock_control_subsys_t) &scc_subsys, NULL);
```
indentation looks off

```c
        err = clock_control_configure(config->clkctrl_dev,                      
                                      (clock_control_subsys_t)&scc_subsys,      
                                      NULL);                                    
```

fabiobaltieri commented on drivers/i2c/i2c_numaker.c:
```diff
@@ -0,0 +1,808 @@
+		LOG_ERR("MSG IDX: %d", data->master_xfer.msgs_pos - data->master_xfer.msgs_beg);
+		LOG_ERR("ADDR (%d-bit): 0x%04X", is_10bit ? 10 : 7, addr);
+		LOG_ERR("DIR: %s", is_read ? "R" : "W");
+		LOG_ERR("Expected %d bytes transferred, but actual %d",
+			 data->master_xfer.msgs_pos->len,
+			 data->master_xfer.buf_pos - data->master_xfer.buf_beg);
```
indentation looks off

```c
                LOG_ERR("Expected %d bytes transferred, but actual %d",         
                        data->master_xfer.msgs_pos->len,                        
                        data->master_xfer.buf_pos - data->master_xfer.buf_beg); 
```

fabiobaltieri commented on drivers/i2c/i2c_numaker.c:
```diff
@@ -0,0 +1,808 @@
+	return 0;
+}
+
+static int i2c_numaker_transfer(const struct device *dev, struct i2c_msg *msgs,
+				 uint8_t num_msgs, uint16_t addr)
```
```c
static int i2c_numaker_transfer(const struct device *dev, struct i2c_msg *msgs, 
                                uint8_t num_msgs, uint16_t addr)                
```

you prob want to check your editor to make it do it for you, there's few odds indentations around the file :-)

cyliangtw replied:
Yes, it's fixed after adjust my editor tab size from 4 as 8.
Thanks of your advice.

fabiobaltieri commented on drivers/i2c/i2c_numaker.c:
```diff
@@ -0,0 +1,808 @@
+                                                                                 \
+	static void i2c_numaker_irq_config_func_##inst(const struct device *dev) \
+	{                                                                        \
+		IRQ_CONNECT(DT_INST_IRQN(inst),                                  \
+					DT_INST_IRQ(inst, priority),             \
+					i2c_numaker_isr,                         \
+					DEVICE_DT_INST_GET(inst),                \
+					0);                                      \
```
```c
                IRQ_CONNECT(DT_INST_IRQN(inst),                                  \
                            DT_INST_IRQ(inst, priority),                         \
                            i2c_numaker_isr,                                     \
                            DEVICE_DT_INST_GET(inst),                            \
                            0);                                                  \
```

fabiobaltieri commented on drivers/i2c/i2c_numaker.c:
```diff
@@ -0,0 +1,808 @@
+                                                                                 \
+	static struct i2c_numaker_data i2c_numaker_data_##inst;                  \
+                                                                                 \
+	I2C_DEVICE_DT_INST_DEFINE(inst,                                          \
+			i2c_numaker_init,                                        \
+			NULL,                                                    \
+			&i2c_numaker_data_##inst,                                \
+			&i2c_numaker_config_##inst,                              \
+			POST_KERNEL,                                             \
+			CONFIG_I2C_INIT_PRIORITY,                                \
+			&i2c_numaker_driver_api);
```
```c
        I2C_DEVICE_DT_INST_DEFINE(inst,                                          \
                                  i2c_numaker_init,                              \   
                                  NULL,                                          \   
                                  &i2c_numaker_data_##inst,                      \   
                                  &i2c_numaker_config_##inst,                    \   
                                  POST_KERNEL,                                   \   
                                  CONFIG_I2C_INIT_PRIORITY,                      \   
                                  &i2c_numaker_driver_api);                     
```

fabiobaltieri commented on drivers/i2c/i2c_numaker.c:
```diff
@@ -0,0 +1,808 @@
+	__ASSERT_NO_MSG((data->master_xfer.buf_end - data->master_xfer.buf_pos) == 1);
+
+	/* Flags of previous message */
+	bool do_stop_prev = (data->master_xfer.msgs_pos->flags & I2C_MSG_STOP);
```
nit: no need for the praenthesis on this one

fabiobaltieri commented on drivers/i2c/i2c_numaker.c:
```diff
@@ -0,0 +1,808 @@
+	if (data->master_xfer.msgs_pos < data->master_xfer.msgs_end) {
+		/* Flags of next message */
+		struct i2c_msg *msgs_pos = data->master_xfer.msgs_pos;
+		bool is_read_next = ((msgs_pos->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ);
+		bool do_restart_next = (data->master_xfer.msgs_pos->flags & I2C_MSG_RESTART);
```
same, no ned for parentehsis on these

cyliangtw replied:
Thanks of you advice and update for above comments at commit:  01e0dd0d54 & 8e423a12421

fabiobaltieri reviewed (commented):
good stuff, few comments, mainly coding style

cyliangtw commented on drivers/i2c/i2c_numaker.c:
```diff
@@ -0,0 +1,808 @@
+#define  ADDR_TRAN_ARB_LOST    0xB0  /* Address Transmit Arbitration Lost */
+#define  BUS_RELEASED          0xF8  /* Bus Released */
+
+/* Implementation notes
+ * 1. Support dual role
+ * 2. Support at most one slave at one time.
+ * 3. Support merging adjacent messages into one transaction.
+ * 4. Force I2C Stop to end the transfer
+ */
+
+/* Transfer control type
+ * Active transfer (master only):
+ * 1. Do I2C Start to start the transfer (thread)
+ * 2. I2C FSM (ISR)
+ * 3. Force I2C Stop to end the transfer (thread)
+ *
+ * Passive transfer (slave only):
+ * 1. Prepare callback (thread)
+ * 2. Do data transfer via above callback (ISR)
+ */
```
OK, I remove it and move the description of transfer to i2c_numaker_transfer().

cyliangtw commented on tests/drivers/i2c/i2c_target_api/boards/numaker_pfm_m467.overlay:
```diff
@@ -0,0 +1,41 @@
+&pinctrl {
+	i2c1_default: i2c1_default {
+		group0 {
+			pinmux = <PB2MFP_I2C1_SDA>,  /* UNO D0 */
+					<PB3MFP_I2C1_SCL>;  /* UNO D1 */
```
Yes and aligned.

cyliangtw committed:
tests: drivers: i2c: i2c_target_api: support numaker_pfm_m467

Add support for Nuvoton numaker board numaker_pfm_m467.

Signed-off-by: cyliang tw <cyliang@nuvoton.com>

fabiobaltieri commented on drivers/i2c/i2c_numaker.c:
```diff
@@ -0,0 +1,792 @@
+	.recover_bus = i2c_numaker_recover_bus,
+};
+
+#define NUMAKER_CLKCTRL_DEV_INIT(inst)                                         \
+	.clkctrl_dev = DEVICE_DT_GET(DT_PARENT(DT_INST_CLOCKS_CTLR(inst))),
+
+#define NUMAKER_PINCTRL_DEFINE(inst)                                           \
+	PINCTRL_DT_INST_DEFINE(inst);
+#define NUMAKER_PINCTRL_INIT(inst)                                             \
+	.pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),
```
are these still needed? I imagine they used to be in a conditional and defined to noop at some point but now you may as well drop them and use the expansion directly... makes it a bit easier to follow

cyliangtw replied:
Yes, it's not need anymore.
Change to the expansion directly at commit fea78870cd3.

cyliangtw committed:
drivers: i2c: support for Nuvoton numaker series

Add Nuvoton numaker series I2C controller feature.
Support dual role and at most one slave at one time

Signed-off-by: cyliang tw <cyliang@nuvoton.com>
